projects/ad2s1210_iio/app/ad2s1210_iio.c

```diff
@@ -0,0 +1,619 @@
+		AD2S1210_CHN_ATTR("raw", RAW_ATTR_ID),
+		AD2S1210_CHN_ATTR("scale", SCALE_ATTR_ID),
+		END_ATTRIBUTES_ARRAY
+	},
+	{
+		AD2S1210_CHN_ATTR("frequency", FREQ_ATTR_ID),
+		AD2S1210_CHN_ATTR("frequency_available", FREQ_AVAIL_ATTR_ID),
+		AD2S1210_CHN_ATTR("label", LABEL_ATTR_ID),
+		END_ATTRIBUTES_ARRAY
+	},
+};
+
+/* IIO device (global) attributes list */
+static struct iio_attribute ad2s1210_iio_global_attributes[] = {
+	AD2S1210_CHN_ATTR("sampling_frequency", SAMPLING_FREQ_ATTR_ID),
+	END_ATTRIBUTES_ARRAY
+};
+
+/* IIO channels info */
+static struct iio_channel ad2s1210_iio_channels[RESOLVER_CHANNELS] = {
+	AD2S1210_CH("position", 0, IIO_ANGL, false),
+	AD2S1210_CH("velocity", 1, IIO_ANGL_VEL, false),
+	AD2S1210_CH("altvoltage", 2, IIO_ALTVOLTAGE, true)
+};
+
+/* Flag to indicate if size of the buffer is updated according to requested
+ * number of samples for the multi-channel IIO buffer data alignment
+ */
+static bool buf_size_updated;
+
+static const float ad2s1210_velocity_scale[] = {
+	17089132, /* 8.192MHz / (2*pi * 2500 / 2^15) */
+	42722830, /* 8.192MHz / (2*pi * 1000 / 2^15) */
+	85445659, /* 8.192MHz / (2*pi * 500 / 2^15) */
+	341782638, /* 8.192MHz / (2*pi * 125 / 2^15) */
+};
+
+static int iio_ad2s1210_attr_get(void *device, char *buf, uint32_t len,
+				 const struct iio_ch_info *channel,
+				 intptr_t priv)
+{
+	uint16_t data = 0;
+	uint16_t data_cpu = 0;
+	struct ad2s1210_dev *dev = (struct ad2s1210_dev *) device;
+	enum ad2s1210_channel chn = AD2S1210_POS;
+	float rps_max;
+	int ret;
+	uint16_t fexcit;
+
+	if (channel->type == IIO_ANGL_VEL)
+		chn = AD2S1210_VEL;
+
+	switch (priv) {
+	case RAW_ATTR_ID:
+		ret = ad2s1210_spi_single_conversion(device, chn, &data);
+		if (ret)
+			return ret;
+
+		data_cpu = swab16(data);
+
+		if (channel->type == IIO_ANGL_VEL)
+			return sprintf(buf, "%d", (int16_t)data_cpu);
+
+		return sprintf(buf, "%u", data_cpu);
+	case SCALE_ATTR_ID:
+		if (channel->type == IIO_ANGL_VEL) {
+			rps_max = ad2s1210_velocity_scale[dev->resolution];
+			return snprintf(buf, len, "%10f", AD2S1210_FCLKIN / rps_max);
+		}
+
+		return snprintf(buf, len, "%10f", AD2S1210_POS_IIO_SCALE);
+
+	case SAMPLING_FREQ_ATTR_ID:
+		return snprintf(buf, len, "%d", SAMPLING_RATE);
+	case LABEL_ATTR_ID:
+		if (channel->type == IIO_ANGL_VEL)
+			return snprintf(buf, len, "velocity");
+		else if (channel->type == IIO_ALTVOLTAGE)
+			return snprintf(buf, len, "excitation");
+		return snprintf(buf, len, "position");
+
+	case HYSTERESIS_ATTR_ID:
+		ret = ad2s1210_hysteresis_is_enabled(dev);
+		if (ret < 0)
+			return ret;
+		return snprintf(buf, len, "%d", ret);
+
+	case HYSTERESIS_AVAILABLE_ATTR_ID:
+		return snprintf(buf, len, "0 1");
+
+	case FREQ_ATTR_ID:
+		ret = ad2s1210_get_excitation_frequency(dev, &fexcit);
+		if (ret)
+			return ret;
+		return snprintf(buf, len, "%hu", fexcit);
+
+	case FREQ_AVAIL_ATTR_ID:
+		return snprintf(buf, len, "[%d %d %d]", AD2S1210_MIN_EXCIT,
+				250, AD2S1210_MAX_EXCIT);
```

Comment:
What is the significance of the number -250?

Reply:
this is needed to be in par with the linux attributes (comes from linux driver). 250 is the step size for each selectable frequency between min and max. i will add a macro to make it clear.